src/client/auth/mod.rs

```diff
@@ -185,6 +192,43 @@ impl AuthMechanism {
         }
     }
 
+    /// Constructs the first message to be sent to the server as part of the authentication
+    /// handshake, which can be used for speculative authentication.
+    pub(crate) fn build_client_first(
```

Comment:
The strategy I took to implement speculative auth was to separate out the building (and sending) of the first round of the SCRAM and X.509 auth handshakes (as they're the only mechanisms that support this feature). To that end, I defined a separate message to build and return the first client message in the handshake so that it can be put in the handshake.